core/src/main/java/org/apache/iceberg/TableMetadata.java

```diff
@@ -502,7 +519,10 @@ public TableMetadata replaceCurrentSnapshot(Snapshot snapshot) {
   }
 
   public TableMetadata removeSnapshotsIf(Predicate<Snapshot> removeIf) {
-    List<Snapshot> toRemove = snapshots.stream().filter(removeIf).collect(Collectors.toList());
+    List<Snapshot> toRemove = snapshots.stream()
+        .filter(removeIf)
+        .filter(snapshot -> !refsById.containsKey(snapshot.snapshotId()))
+        .collect(Collectors.toList());
```

Comment:
I think it is probably time to remove this method, or at least to stop using it. Instead, I'd probably just produce a list of snapshot IDs to remove. That fits better with the new builder style.

Reply:
added https://github.com/apache/iceberg/issues/3900 to track this